### DISCUSSION TITLE
G1 - Cadenas dinamicas

### DISCUSSION THREAD

**Discussion by Nikzza** [INGCOM-UNRN-P1/cursada-2023#191] [Preguntas y respuestas] [answered]
tengo una consulta 
![image](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/141664052/fbf8d1d1-8400-4b5c-81cb-47011f232006)

mi consulta es la primera funcion se refiere a pedir literalmente caracteres por separado? o pedir por ejemplo "Hola"
y la otra funcion te devuelve los caracteres que pediste.
Pero no faltaria una funcion que te pida que cantidad de capacidad queres usar del arreglo?

**Comment by martinvilu** [comment]
> mi consulta es la primera funcion se refiere a pedir literalmente caracteres por separado? o pedir por ejemplo "Hola"

Ninguna de las dos opciones, `pedir` es como un `malloc`, para pedir un bloque de caracteres. Potencialmente, dejando un `\0` al principio de lo retornado.

> y la otra funcion te devuelve los caracteres que pediste.

Sí, los deja disponibles para volver a pedirlos

> Pero no faltaria una funcion que te pida que cantidad de capacidad queres usar del arreglo?

No, esto es dinámico e individual al usar la función de pedir. Pero no termino de entender tu planteo. Así que, ¿porque no me explicas un poco más a qué te referís?

**Reply by Nikzza** [reply]
lo que yo pense es que seria como una funcion en la que vos pedis una cierta capacidad para poder llenarla de caracteres, ese "bloque" que vos pediste podes tanto rellenarlo o como dejarlo vacio pero es tu bloque. por eso decia si tenia que pedir la capacidad.

**Reply by Nikzza** [reply]
no se si comprendi al 100 el ejercicio por eso las preguntas, creo que no termine de entender la finalidad

**Reply by Nikzza** [reply]
yo porque como hay un ejemplo que das que el usuario pide
Hola (capacidad 8, largo 4)
Mundo (capacidad 6, largo 5)
P1! (capacidad 5, largo 3)

yo pienso que capacidad 8 es el bloque que pidio de 8 bytes y solo utilizó 4 de ellos, el resto quedó vacio.
Por lo que entiendo yo el pedido de memoria del malloc se basa en la capacidad que pidio el usuario y no en el largo de la cadena

**Reply by martinvilu** [reply]
Son tres llamadas a pedir, una con 8, después con 6, y 5.

A esas cadenas les copias o cargas caracteres.

Olvidate de los bytes, este es un `malloc` simplificado que solo maneja de a caracteres. (mas adelante se puede pensar en algo mas parecido, usando punteros genéricos en lugar de a `char`).

**Comment by martinvilu** [comment, marked as answer]
> lo que yo pense es que seria como una funcion en la que vos pedis una cierta capacidad para poder llenarla de caracteres, ese "bloque" que vos pediste podes tanto rellenarlo o como dejarlo vacio pero es tu bloque. por eso decia si tenia que pedir la capacidad.

Si, `caracteres` hace referencia a cuantos caracteres pedís para el bloque; su capacidad.

**Reply by Nikzza** [reply]
esta bien
dejame lo termino de acomodar y te mando lo que tengo para ver si estoy bien encaminado

**Reply by Nikzza** [reply]
si no es molestia, la verdad q esta bueno el ejercicio asi q me voy a quedar un rato

**Reply by martinvilu** [reply]
Ninguna molestia, es el mejor ejercicio para entender el funcionamiento de `malloc` :-)

**Reply by martinvilu** [reply]
Siendo que son cadena, hay un conjunto de cosas para hacer que son muy interesantes:
* creá un `main` en el que la pileta inicial no sea demasiado grande, 
* Pedí varias cadenas, como para que la pileta se agote
* Libera alguna cadena (del medio)
* Implementá una función que te permita ver los caracteres 'dueños' de la pileta (o incluso enviarla a un archivo)

Lo más interesante del ejercicio está cuando luego de liberar algo 'del medio' ¿que debiera de pasar? ¿Realmente podés decir que no hay más lugar? ¿Cómo buscas un espacio libre?, si recién se liberó

**Reply by Nikzza** [reply]
lo que hice fue pedir el espacio de memoria osea los caracteres aunque no coloque ninguna cadena el espacio ya esta asignado no? por mas que no ingrese nada.

**Reply by Nikzza** [reply]
si libero el espacio de alguna cadena del medio deberia estar disponible como para poder pedir una cadena del mismo tamaño o menor.
No podes decir que no hay mas lugar ya que la memoria se libero, la capacidad de la pileta depende de como se pidieron y liberaron las cadenas
No lo podrias buscar desde el punto en donde se libero la pileta? podes fijarte que posiciones hay disponible en la pileta

**Comment by Nikzza** [comment]
Ahora tengo un problema, necesito pasar el struct a las funciones para que puedan hacer las operaciones necesarias, no se como pasarla a otro modulo, ya que si tengo 2 funciones una requiere de la otra y al poner el include de cada modulo en las 2 se bugea todo

**Reply by martinvilu** [reply]
Ahi hacen falta headers y opciones de compilacion mas completas

**Reply by Nikzza** [reply]
ya lo pude hacer, hice un header en la que esta la estructura para poder usarla sin q se bugee todo

**Comment by martinvilu** [comment]
Estoy viendo que a las funciones, les faltan la pileta para estar completas. De otra forma, no sabes de dónde salió la memoria para ampliar,  devolver y consultar.


```C
char* pedir(pileta_t *pile, int caracteres);
int devolver(pileta_t* pile, char* cadena);
int redimensionar(pileta_t* pile, char* cadena, int nuevo);
int capacidad(pileta_t* pile, char* cadena);
```

**Reply by Nikzza** [reply]
te referis al pedazo de pileta que se pidio?

**Reply by martinvilu** [reply]
Sin el puntero (en `pedir`, en particular), la pileta que pasas es una copia :-) y la memoria que retornas es relativa a esa copia en el stack. Por lo que el puntero devuelto queda en un lugar incorrecto.

**Reply by Nikzza** [reply]
ahhhh
